x/incentives/keeper/distribution.go

```diff
@@ -0,0 +1,171 @@
+package keeper
+
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/tharsis/evmos/x/incentives/types"
+)
+
+// Distribute transfers the allocated rewards to the participants of a given
+// incentive.
+//  - allocates the amount to be distribbuted from the inflaction pool
+//  - distributes the rewards to all paricpants
+//  - deletes all gas meters
+//  - updates the remaining epochs of each incentive
+//  - sets the cumulative totalGas to zero
+func (k Keeper) DistributeIncentives(ctx sdk.Context) error {
+	logger := k.Logger(ctx)
+
+	// allocate rewards for each Incentive
+	coinsAllocated := k.allocateCoins(ctx)
+
+	k.IterateIncentives(
+		ctx,
+		func(incentive types.Incentive) (stop bool) {
+			// distribute rewards
+			k.rewardParticipants(ctx, incentive, coinsAllocated)
+
+			// update epoch and remove incentive from epoch if already finalized
+			incentive.Epochs--
+			if !incentive.IsActive() {
+				k.DeleteIncentive(ctx, incentive)
+				logger.Info(
+					"incentive finalized",
+					"contract", incentive.Contract,
+				)
+			} else {
+				k.SetIncentive(ctx, incentive)
+
+				// reset incentive's total gas count
+				k.SetIncentiveTotalGas(ctx, incentive, 0)
+			}
+
+			ctx.EventManager().EmitEvent(
+				sdk.NewEvent(
+					types.EventTypeDistributeIncentives,
+					sdk.NewAttribute(types.AttributeKeyContract, incentive.Contract),
+					sdk.NewAttribute(
+						types.AttributeKeyEpochs,
+						fmt.Sprint(incentive.Epochs),
+					),
+				),
+			)
+			return false
+		})
+
+	return nil
+}
+
+// Allocate amount of coins to be distributed for each incentive
```

Comment:
I think this requires a more descriptive comment. 

- Iterate over all the registered and active incentives
- create an allocation (module account) from escrow balance to be distributed to the contract address
- checks that escrow balance is sufficient